File: Session 1/code/network-java/src/main/java/com/compnet/practical1/A5ProducerKO.java
/*
 * 
 */

package com.compnet.practical1;

import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 *
 * @author 
 */
public class A5ProducerKO {
    
    public static void main(String[] args) throws InterruptedException {
        
        Display d = new Display();

        sleep(500, 0);
        int n = 10;
        for (int i = 0; i < n; i++) {
            sleep(0, (int) (Math.random()*999999));
            producer(d).start();
        }
    }

    private static Thread producer(Display d) {
        return new Thread(() -> {
            for (int i = 0; i < 100; i++) {
                int ind = (int) (Math.random()*20);
                d.add_value(ind, 1);
                sleep((int) (Math.random()*2), 0);
                d.add_value(ind, -1);
            }
            d.render_clean();
        });
    }
        
    private static void sleep(long ms, int nanos) {
        try {
            Thread.sleep(ms, nanos);
        } catch (InterruptedException ex) {
            Logger.getLogger(Display.class.getName()).log(Level.SEVERE, null, ex);
        }
    }    

}

File: Session 1/code/network-python/A5_producer_ko.py

from threading import Thread
import time
import random
import display

def main(): # called at the end of the file
    d = display.Display()

    time.sleep(0.5)
    n = 10
    for i in range(n):
        time.sleep(random.uniform(0, 0.001))
        producer(d).start()

def producer(d):
    def add_elements():
        for iteration in range(100):
            i = random.randrange(20)
            d.add_value(i, 1)
            time.sleep(random.uniform(0, 0.002))
            d.add_value(i, -1)
        d.render_clean()

    t = Thread(target=add_elements)
    return t

if __name__== "__main__":
  main()
